lib/remoteproc/remoteproc_virtio.c

```diff
@@ -111,20 +111,37 @@ static uint32_t rproc_virtio_negotiate_features(struct virtio_device *vdev,
 static void rproc_virtio_read_config(struct virtio_device *vdev,
 				     uint32_t offset, void *dst, int length)
 {
-	(void)vdev;
-	(void)offset;
-	(void)dst;
-	(void)length;
+	struct remoteproc_virtio *rpvdev;
+	struct fw_rsc_vdev *vdev_rsc;
+	struct metal_io_region *io;
+	char *config;
+
+	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
+	vdev_rsc = rpvdev->vdev_rsc;
+	config = (char *)(&vdev_rsc->vring[vdev->vrings_num]);
+	io = rpvdev->vdev_rsc_io;
+	metal_io_block_read(io,
+			    metal_io_virt_to_offset(io, config + offset),
+			    dst, length);
 }
 
 #ifndef VIRTIO_SLAVE_ONLY
 static void rproc_virtio_write_config(struct virtio_device *vdev,
 				      uint32_t offset, void *src, int length)
 {
-	(void)vdev;
-	(void)offset;
-	(void)src;
-	(void)length;
+	struct remoteproc_virtio *rpvdev;
+	struct fw_rsc_vdev *vdev_rsc;
+	struct metal_io_region *io;
+	char *config;
+
+	rpvdev = metal_container_of(vdev, struct remoteproc_virtio, vdev);
+	vdev_rsc = rpvdev->vdev_rsc;
+	config = (char *)(&vdev_rsc->vring[vdev->vrings_num]);
```

Comment:
Need to check the config_len first.  Please, check that config_len is not null but also that the requested size + offset is  not higher than the config_len.

Reply:
Sure, I will update in the next pull request.

Reply:
looks like new remoteproc_virtio variable is required to keep the config len, can we always assume that the config space is always right after the predefined vdev space if the shared buffer feature is set?
Also, just thinking if we want to make sure offset is 32bit aligned.

Reply:
> > looks like new remoteproc_virtio variable is required to keep the config len,
> 
> Yes.
> 
> > can we always assume that the config space is always right after the predefined vdev space if the shared buffer feature is set?
> 
> From remoteproc.h
> /**
> 
>     * struct fw_rsc_vdev - virtio device header
> 
>     * @config_len is the size of the virtio config space of this vdev. The config
> 
>     * space lies in the resource table immediate after this vdev header.
> 
>     * Note: immediately following this structure is the virtio config space for
> 
>     * this vdev (which is specific to the vdev; for more info, read the virtio
> 
>     * spec). the size of the config space is specified by @config_len.
>       */
> 
> 
> As far as I know, there isn't config defined yet, so the buffer size feature should be first one.

As the config space can already be used for customized configuration. And thus, I want to see if everyone is OK to assume that once RPMsg shared buffer size feature bit is set, we assume that the first offset is for the RPMsg shared buffer size.

Or we make some virtio config structure like 
uint32 num_config_objs;
array = {
{
uint32 config_obj_type;
uint32 offset;
}
...
}
followed by config objs

This will make the config space more complicated. But maybe more flexible even for future if we want to add more "common" configurable objects.

> 
> > Also, just thinking if we want to make sure offset is 32bit aligned.
> 
> What do you want to point out?

Sorry that I didn't clarify what I meant. I was thinking more like "common config space alignment".
assuming that the first N bytes are are for "common" predefined configuration (This is for future features extension), which can be followed by user defined configuration. 32bit is not enough.
The feature you are adding is an interesting feature, I don't want to over complicate things. but my concern is for those who are already using the "config" space for their own platform. Any new common "config" we add, they will need to change their resource table. And thus, want to see if we also want to make agreement on the virtio config space with this patch.

Reply:
> > > looks like new remoteproc_virtio variable is required to keep the config len,
> > 
> > 
> > Yes.
> > > can we always assume that the config space is always right after the predefined vdev space if the shared buffer feature is set?
> > 
> > 
> > From remoteproc.h
> > /**
> > ```
> > * struct fw_rsc_vdev - virtio device header
> > 
> > * @config_len is the size of the virtio config space of this vdev. The config
> > 
> > * space lies in the resource table immediate after this vdev header.
> > 
> > * Note: immediately following this structure is the virtio config space for
> > 
> > * this vdev (which is specific to the vdev; for more info, read the virtio
> > 
> > * spec). the size of the config space is specified by @config_len.
> >   */
> > ```
> > As far as I know, there isn't config defined yet, so the buffer size feature should be first one.
> 
> As the config space can already be used for customized configuration. And thus, I want to see if everyone is OK to assume that once RPMsg shared buffer size feature bit is set, we assume that the first offset is for the RPMsg shared buffer size.

1.Config space concept doesn't expose to rpmsg driver API level
2.No virtio based backend(e.g. QCOM SMD) doesn't have this concept too
so rpmsg driver(client) is hard(or should avoid) to utilize this feature, and reserve the config spase to the internal implementation?
If somebody use the config space internally, since:
1.They don't share the information publicly
2.The internal field may not general enough to other people
It's better to put the general information at the first, then the optional private field last.

> 
> Or we make some virtio config structure like
> uint32 num_config_objs;
> array = {
> {
> uint32 config_obj_type;
> uint32 offset;
> }
> ...
> }
> followed by config objs
> 
> This will make the config space more complicated. But maybe more flexible even for future if we want to add more "common" configurable objects.
> 

This approach will complicate the implementation and also require the change for current config space user.

> > > Also, just thinking if we want to make sure offset is 32bit aligned.
> > 
> > 
> > What do you want to point out?
> 
> Sorry that I didn't clarify what I meant. I was thinking more like "common config space alignment".
> assuming that the first N bytes are are for "common" predefined configuration (This is for future features extension), which can be followed by user defined configuration. 32bit is not enough.
> The feature you are adding is an interesting feature, I don't want to over complicate things. but my concern is for those who are already using the "config" space for their own platform. Any new common "config" we add, they will need to change their resource table. And thus, want to see if we also want to make agreement on the virtio config space with this patch.

How about this:
struct virtio_rpmsg_config {
  __u32 txbuf_size;
  __u32 rxbuf_size;
  __u32 reserved[14];
}; __attribute__((packed));
so
1.The standard feature could use the reserved field in the furture without breaking compatiblity
2.The internal config could use the space after reserved field
3.Keep the implementation simple